src/riot-rs-random/src/lib.rs

```diff
@@ -0,0 +1,117 @@
+//! This is taken care of by the `riot-rs-embassy` initialization functions. Applications can
+//! ensure that this has happened by depending on the laze feature `rng`.
+//!
+//! ---
+//!
+//! Currently, this provides very little choice, and little fanciness: It (more or less
+//! arbitrarily) uses the [rand_chacha::ChaCha20Rng] generator, seeds uses different copies of that
+//! RNG for all callers fo [`get_rng()`]. Later, it may offer a wider range of choices and
+//! trade-offs between flash size, speed and security. (See [Rng] for its properties).
+#![no_std]
+
+use rand_core::{RngCore, SeedableRng};
+
+// This is one of the points where we can tune easily. If calls to `get_rng()` are rare, it may
+// even make sense to move the HWRNG in here to get a global ZST.
+// #[cfg(feature = "main-is-csprng")]
+pub(crate) type SelectedRng = rand_chacha::ChaCha20Rng;
+
+/// The OS provided random number generator.
+///
+/// This may be small, fast and/or secure, but will sacrifice the earlier properties if the later
+/// are needed by any system component.
+///
+/// Such an RNG can be requested by any component, and will always be seeded appropriately.
+///
+/// There is no point in requesting a "small" RNG (if a fast RNG is built in, that will do just as
+/// well). There may be a point in requesting a "fast" RNG, which it may make sense to introduce a
+/// FastRng that particular components can request. (Until that is available, an application can
+/// depend on a known fast RNG and seed it from the OS's RNG, but preferably there should be a
+/// single configured implementation used across applications for ROM optimization, even if it
+/// turns out to be beneficial to have a thread or task local RNG around in RAM for speed).
+pub struct Rng {
+    inner: SelectedRng,
+}
+
+impl Rng {
+    // Create a per-thread/per-task copy of the RNG. Not sure whether we'll need that all the time,
+    // and/or whether these should all have the same types, but for the first iteration
+    // everything-is-of-type-Rng is good enough
+    fn split(&mut self) -> Self {
+        Self { inner: SelectedRng::from_rng(self).expect("Main RNG is infallible") }
+    }
+}
+
+// Re-implementing the trait rather than Deref'ing into inner: This avoids leaking implementation
+// details to users who might otherwise come to depend on platform specifics of the Rng.
+impl RngCore for Rng {
+    fn next_u32(&mut self) -> u32 {
+        self.inner.next_u32()
+    }
+    fn next_u64(&mut self) -> u64 {
+        self.inner.next_u64()
+    }
+    fn fill_bytes(&mut self, buf: &mut [u8]) {
+        self.inner.fill_bytes(buf)
+    }
+    fn try_fill_bytes(&mut self, buf: &mut [u8]) -> Result<(), rand_core::Error> {
+        self.inner.try_fill_bytes(buf)
+    }
+}
+
+#[cfg(feature = "main-is-csprng")]
+mod main_is_csprng {
+    impl rand_core::CryptoRng for super::Rng {}
+
+    fn generate_inner_type() -> super::SelectedRng {
```

Comment:
I don't really understand what these two functions do.

Reply:
Will be documented: These typecheck that indeed SelectedRng implements CryptoRng. (If we had something like `#[derive(CryptoRng)]`, that'd be checked automatically, but as we just add an impl, I want this typechecked).

To be improved in the docs.